lib/ask/runtime/questionnaire_action.ex

```diff
@@ -0,0 +1,245 @@
+defmodule Ask.Runtime.QuestionnaireAction do
+  alias Ask.Runtime.QuestionnaireExport
+
+  def export(questionnaire) do
+    QuestionnaireExport.export(questionnaire)
+  end
+end
+
+defmodule Ask.Runtime.QuestionnaireExport do
+  alias Ask.{Questionnaire, Repo, Audio}
+
+  def export(questionnaire) do
+    questionnaire = clean_i18n_quiz(questionnaire)
+    all_questionnaire_steps = Questionnaire.all_steps(questionnaire)
+    audio_ids = collect_steps_audio_ids(all_questionnaire_steps, [])
+    audio_ids = collect_settings_audio_ids(questionnaire.settings, audio_ids)
+    # for each audio: charges it in memory and then streams it.
+    audio_resource =
+      Stream.resource(
+        fn -> audio_ids end,
+        fn audio_ids ->
+          case audio_ids do
+            [id | tail] ->
+              audio = Repo.get_by(Audio, uuid: id)
+              {[audio], tail}
+
+            [] ->
+              {:halt, audio_ids}
+          end
+        end,
+        fn _ -> [] end
+      )
+
+    # The audio file import is based on this list
+    # If it's empty, nothing will be imported
+    audio_files =
+      Stream.map(audio_resource, fn audio ->
+        %{
+          "uuid" => audio.uuid,
+          "original_filename" => audio.filename,
+          "source" => audio.source
+        }
+      end)
+
+    audio_entries =
+      Stream.map(audio_resource, fn audio ->
+        # Zstream needs to recieve audio.data as enumerable in order to work, otherwise it throws Protocol.undefined error.
+        Zstream.entry("audios/" <> Audio.exported_audio_file_name(audio.uuid), [audio.data])
+      end)
+
+    manifest = %{
+      name: questionnaire.name,
+      modes: questionnaire.modes,
+      steps: questionnaire.steps,
+      quota_completed_steps: questionnaire.quota_completed_steps,
+      settings: questionnaire.settings,
+      partial_relevant_config: questionnaire.partial_relevant_config,
+      languages: questionnaire.languages,
+      default_language: questionnaire.default_language,
+      audio_files: audio_files
+    }
+
+    {:ok, json} = Poison.encode(manifest)
+
+    json_entry =
+      Stream.map([json], fn json ->
+        Zstream.entry("manifest.json", [json])
+      end)
+
+    zip_entries = Stream.concat(audio_entries, json_entry)
+    # since audio binary data is created as list for Zstream
+    # flatten is needed to allow the data to be sent in chunks
+    # otherwise the connection sends the whole list as a chunk
+    # and times out.
+    Zstream.zip(zip_entries)
+    |> Stream.flat_map(fn element ->
+      case is_list(element) do
+        true -> List.flatten(element)
+        false -> element
+      end
+    end)
+  end
+
+  def clean_i18n_quiz(quiz) do
+    clean_i18n_fields = Map.keys(quiz)
+
+    Enum.reduce(clean_i18n_fields, quiz, fn field, quiz_acc ->
```

Comment:
Why we need to iterate fields this way?

Can't we be more explicit and simple like:

```elixir
quiz |> clean(:setting, "error_message")
     |> clean(:setting, "thank_you_message")
     |> clean(:steps, ".[].prompt")
     |> clean(:steps, ".[].choices.[].responses.[]")
     |> clean(:steps, ".[].refusal.responses")
     |> clean(:quota_completed_steps, ".[].prompt")
     |> clean(:quota_completed_steps, ".[].choices.[].responses.[]")
     |> clean(:quota_completed_steps, ".[].refusal.responses")
```

Also, why is `error_message` wihout a preceding dot, but `.[]` needs one? I would expect to be `.error_message`